maven-core/src/main/java/org/apache/maven/project/MavenProject.java

```diff
@@ -921,12 +926,20 @@ public void setInjectedProfileIds( String source, List<String> injectedProfileId
      * coordinates.
      *
      * @param artifact the artifact to add or replace.
-     * @throws DuplicateArtifactAttachmentException
+     * @deprecated Please use {@link MavenProjectHelper}
+     * @throws DuplicateArtifactAttachmentException not used anymore but leave it for backward compatibility
      */
     public void addAttachedArtifact( Artifact artifact )
         throws DuplicateArtifactAttachmentException
     {
-        getAttachedArtifacts().add( artifact );
+        // if already there we remove it and add again
+        int index = attachedArtifacts.indexOf( artifact );
```

Comment:
This will incur a complexity of `O(n)`. Would it make sense to work with a set/map to make that test?

Reply:
we are talking about list of 6/7 elements which are not accessed that much especially for add.
My first idea was to use Set but this will make the `getAttachedArtifacts` method to recreate a List based on the  Set and makes this new List unmodifiable as well.
I can make some jmh to compare but not sure it really worth it...

Reply:
No, it is not worth it. If the amount of elements is low, I am fine with indexed list access.